pkg/unikernel/service/container.go

```diff
@@ -0,0 +1,366 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/golang/glog"
+	libvirtgo "github.com/libvirt/libvirt-go"
+	"k8s.io/frakti/pkg/unikernel/metadata"
+	"k8s.io/frakti/pkg/unikernel/metadata/store"
+
+	kubeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
+)
+
+// CreateContainer creates a new container in specified PodSandbox
+func (u *UnikernelRuntime) CreateContainer(podSandboxID string, config *kubeapi.ContainerConfig, sandboxConfig *kubeapi.PodSandboxConfig) (string, error) {
+	var err error
+	// Check if there is any pod already created in pod
+	// We now only support one-container-per-pod
+	createdContainers, err := u.getAllContainersInPod(podSandboxID)
+	if err != nil {
+		return "", fmt.Errorf("get containers for pod(%q) failed: %v", podSandboxID, err)
+	}
+	for _, container := range createdContainers {
+		glog.Warningf("Unikernel/CreateContainer: container(%q) already exist in pod(%q), remove it", container.ID, podSandboxID)
+		if err := u.RemoveContainer(container.ID); err != nil {
+			glog.Errorf("Clean up legacy container(%q) failed: %v", container.ID, err)
+		}
+	}
+
+	sandbox, err := u.sandboxStore.Get(podSandboxID)
+	if err != nil {
+		return "", fmt.Errorf("failed to get sandbox(%q) from store: %v", podSandboxID, err)
+	}
+
+	cid := generateID()
+	cName := makeContainerName(config.GetMetadata(), sandboxConfig.GetMetadata())
+	if err = u.containerNameIndex.Reserve(cName, cid); err != nil {
+		return "", fmt.Errorf("reserve container name %q failed: %v", cName, err)
+	}
+	defer func() {
+		if err != nil {
+			u.containerNameIndex.ReleaseByName(cName)
+		}
+	}()
+
+	// Create internal container metadata
+	meta := metadata.ContainerMetadata{
+		ID:        cid,
+		Name:      cName,
+		SandboxID: podSandboxID,
+		Config:    config,
+	}
+
+	// TODO(Crazykev): Prepare container image
+	// use a temp image for test
+	imageRef, err := u.prepareTestImage()
```

Comment:
this is not a test file, do not use test funcs

Reply:
Yep, we really need image management here.